### ISSUE TITLE
BlockingConnection hearbeat

### ISSUE THREAD

**Issue by fokhunov** [pika/pika#1053]
Hello, I need your help. Cannot figure out solution.
I'm using Rabbitmq docker container with my python app.

About 10 min after running I having exception in my app: 
```
pika.exceptions.ConnectionClosed: (-1, 'EOF')
```

Following log in rabbitmq docker:
```
2018-05-27 11:05:25.296 [warning] <0.1795.0> closing AMQP connection <0.1795.0> (172.18.0.1:58040 -> 172.18.0.2:5672):
missed heartbeats from client, timeout: 60s
```

Here is my producer code:

```
class Producer(object):
    EXCHANGE_TYPE = "direct"
    EXCHANGE_DURABLE = True

    def __init__(self, url, exchange):
        self._connection = pika.BlockingConnection(pika.URLParameters(url))
        self._channel = self._connection.channel()

        self._exchange = exchange
        self._channel.exchange_declare(
            exchange=self._exchange,
            exchange_type=self.EXCHANGE_TYPE,
            durable=self.EXCHANGE_DURABLE
        )

    def publish(self, route, data):
        self._channel.basic_publish(exchange=self._exchange, routing_key=route, body=data)

    def close(self):
            self._channel.close()
            self._connection.close()

```

url =  'amqp://guest:guest@'+ args.rmq +':5672/%2F?heartbeat=0'
exchange = "my_exchange"

**Comment by vitaly-krugl** [comment]
@fokhunov, which version of Pika are you using in this issue? Please attach DEBUG-level Pika logs from this issue beginning with connection establishment until `ConnectionClosed`.

**Comment by fokhunov** [comment]
Hello @vitaly-krugl, thanks for reply. 

As I can see from logs there is _Creating a HeartbeatChecker: 60_ and RabbitMQ closes connection after 3 minutes. My script tries to publish in RabbitMQ and Exception throwing that connection closed.

Pika version: 
`__version__ = '0.11.2'`

Here is RabbitMq container:
```
2018-05-28 21:47:09.408 [info] <0.3732.0> accepting AMQP connection <0.3732.0> (172.18.0.1:58404 -> 172.18.0.2:5672)
2018-05-28 21:47:09.425 [info] <0.3732.0> connection <0.3732.0> (172.18.0.1:58404 -> 172.18.0.2:5672): user 'guest' authenticated and granted access to vhost '/'
2018-05-28 21:50:07.698 [warning] <0.3732.0> closing AMQP connection <0.3732.0> (172.18.0.1:58404 -> 172.18.0.2:5672):
missed heartbeats from client, timeout: 60s
``` 

Here is my script output:
```
__________________________________________
Starting service Mon May 28 23:47:01 2018
INFO:pika.adapters.base_connection:Connecting to ::1:5672
DEBUG:pika.callback:Processing 0:Connection.Start
DEBUG:pika.callback:Processing use of oneshot callback
DEBUG:pika.callback:0 registered uses left
DEBUG:pika.callback:Removing callback #0: {'callback': <bound method SelectConnection._on_connection_start of <SelectConnection PROTOCOL socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
DEBUG:pika.callback:Calling <bound method SelectConnection._on_connection_start of <SelectConnection PROTOCOL socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>> for "0:Connection.Start"
DEBUG:pika.callback:Added: {'callback': <bound method SelectConnection._on_connection_tune of <SelectConnection START socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG:pika.callback:Processing 0:Connection.Tune
DEBUG:pika.callback:Processing use of oneshot callback
DEBUG:pika.callback:0 registered uses left
DEBUG:pika.callback:Removing callback #0: {'callback': <bound method SelectConnection._on_connection_tune of <SelectConnection START socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
DEBUG:pika.callback:Calling <bound method SelectConnection._on_connection_tune of <SelectConnection START socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>> for "0:Connection.Tune"
DEBUG:pika.connection:Creating a HeartbeatChecker: 60
DEBUG:pika.adapters.select_connection:add_timeout: added timeout -6377231062257697481; deadline=60 at 1527544081.27
DEBUG:pika.callback:Added: {'callback': <bound method SelectConnection._on_connection_open of <SelectConnection TUNE socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG:pika.callback:Processing 0:Connection.OpenOk
DEBUG:pika.callback:Processing use of oneshot callback
DEBUG:pika.callback:0 registered uses left
DEBUG:pika.callback:Removing callback #0: {'callback': <bound method SelectConnection._on_connection_open of <SelectConnection TUNE socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
DEBUG:pika.callback:Calling <bound method SelectConnection._on_connection_open of <SelectConnection TUNE socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>> for "0:Connection.OpenOk"
DEBUG:pika.callback:Processing 0:_on_connection_open
DEBUG:pika.callback:Calling <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x10c68ce60>> for "0:_on_connection_open"
DEBUG:pika.connection:Creating channel 1
DEBUG:pika.callback:Added: {'callback': <bound method SelectConnection._on_channel_cleanup of <SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'only': <Channel number=1 CLOSED conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG:pika.callback:Added: {'callback': <bound method Channel._on_getempty of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'only': None, 'one_shot': False, 'arguments': None}
DEBUG:pika.callback:Added: {'callback': <bound method Channel._on_cancel of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'only': None, 'one_shot': False, 'arguments': None}
DEBUG:pika.callback:Added: {'callback': <bound method Channel._on_flow of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'only': None, 'one_shot': False, 'arguments': None}
DEBUG:pika.callback:Added: {'callback': <bound method Channel._on_close of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG:pika.channel:Entering blocking state on frame <Channel.Open(['out_of_band='])>; acceptable_replies=[<class 'pika.spec.OpenOk'>]
DEBUG:pika.channel:Adding on_synchronous_complete callback
DEBUG:pika.callback:Added: {'callback': <bound method Channel._on_synchronous_complete of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG:pika.channel:Adding passed-in callback
DEBUG:pika.callback:Added: {'callback': <bound method Channel._on_openok of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG:pika.callback:Added: {'callback': <bound method BlockingChannel._on_consumer_cancelled_by_broker of <BlockingChannel impl=<Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>>, 'only': None, 'one_shot': False, 'arguments': None}
DEBUG:pika.callback:Added: {'callback': <bound method _CallbackResult.signal_once of <pika.adapters.blocking_connection._CallbackResult object at 0x10c896518>>, 'only': None, 'one_shot': False, 'arguments': None}
DEBUG:pika.callback:Added: {'callback': <bound method BlockingChannel._on_channel_closed of <BlockingChannel impl=<Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG:pika.callback:Added: {'callback': <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x10c8965a8>>, 'only': None, 'one_shot': False, 'arguments': None}
INFO:pika.adapters.blocking_connection:Created channel=1
DEBUG:pika.callback:Processing 1:Channel.OpenOk
DEBUG:pika.callback:Processing use of oneshot callback
DEBUG:pika.callback:0 registered uses left
DEBUG:pika.callback:Removing callback #0: {'callback': <bound method Channel._on_synchronous_complete of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
DEBUG:pika.callback:Processing use of oneshot callback
DEBUG:pika.callback:0 registered uses left
DEBUG:pika.callback:Removing callback #0: {'callback': <bound method Channel._on_openok of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
DEBUG:pika.callback:Calling <bound method Channel._on_synchronous_complete of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>> for "1:Channel.OpenOk"
DEBUG:pika.channel:0 blocked frames
DEBUG:pika.callback:Calling <bound method Channel._on_openok of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>> for "1:Channel.OpenOk"
DEBUG:pika.channel:Entering blocking state on frame <Exchange.Declare(['arguments={}', 'auto_delete=False', 'durable=True', 'exchange=erp', 'internal=False', 'nowait=False', 'passive=False', 'ticket=0', 'type=direct'])>; acceptable_replies=[<class 'pika.spec.DeclareOk'>]
DEBUG:pika.channel:Adding on_synchronous_complete callback
DEBUG:pika.callback:Added: {'callback': <bound method Channel._on_synchronous_complete of <Channel number=1 OPEN conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG:pika.channel:Adding passed-in callback
DEBUG:pika.callback:Added: {'callback': <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x10c896ab8>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG:pika.callback:Processing 1:Exchange.DeclareOk
DEBUG:pika.callback:Processing use of oneshot callback
DEBUG:pika.callback:0 registered uses left
DEBUG:pika.callback:Removing callback #0: {'callback': <bound method Channel._on_synchronous_complete of <Channel number=1 OPEN conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
DEBUG:pika.callback:Processing use of oneshot callback
DEBUG:pika.callback:0 registered uses left
DEBUG:pika.callback:Removing callback #0: {'callback': <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x10c896ab8>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
DEBUG:pika.callback:Calling <bound method Channel._on_synchronous_complete of <Channel number=1 OPEN conn=<SelectConnection OPEN socket=('::1', 55278, 0, 0)->('::1', 5672, 0, 0) params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>> for "1:Exchange.DeclareOk"
DEBUG:pika.channel:0 blocked frames
DEBUG:pika.callback:Calling <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x10c896ab8>> for "1:Exchange.DeclareOk"
DEBUG:pika.heartbeat:Received heartbeat frame
DEBUG:pika.heartbeat:Received heartbeat frame
DEBUG:pika.heartbeat:Received 2 heartbeat frames, sent 0
DEBUG:pika.heartbeat:Sending heartbeat frame
DEBUG:pika.adapters.select_connection:add_timeout: added timeout -4136541231859227008; deadline=60 at 1527544408.75
ERROR:pika.adapters.base_connection:Read empty data, calling disconnect
INFO:pika.connection:Disconnected from RabbitMQ at localhost:5672 (-1): EOF
DEBUG:pika.heartbeat:Removing timeout for next heartbeat interval
DEBUG:pika.adapters.select_connection:remove_timeout: removed -4136541231859227008
DEBUG:pika.channel:Handling meta-close on <Channel number=1 OPEN conn=<SelectConnection CLOSED socket=None params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>
DEBUG:pika.callback:Processing 1:_on_channel_cleanup
DEBUG:pika.callback:Processing use of oneshot callback
DEBUG:pika.callback:0 registered uses left
DEBUG:pika.callback:Removing callback #0: {'callback': <bound method SelectConnection._on_channel_cleanup of <SelectConnection CLOSED socket=None params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'only': <Channel number=1 CLOSED conn=<SelectConnection CLOSED socket=None params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'one_shot': True, 'arguments': None, 'calls': 0}
DEBUG:pika.callback:Calling <bound method SelectConnection._on_channel_cleanup of <SelectConnection CLOSED socket=None params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>> for "1:_on_channel_cleanup"
DEBUG:pika.connection:Removed channel 1
DEBUG:pika.callback:Clearing out '1' from the stack
DEBUG:pika.callback:Processing 0:_on_connection_closed
DEBUG:pika.callback:Calling <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x10c822cb0>> for "0:_on_connection_closed"
DEBUG:pika.callback:Added: {'callback': <bound method SelectConnection._on_connection_start of <SelectConnection CLOSED socket=None params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG:pika.callback:Incremented callback reference counter: {'callback': <bound method SelectConnection._on_connection_close of <SelectConnection CLOSED socket=None params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 2}
ERROR:pika.adapters.blocking_connection:Connection close detected; result=BlockingConnection__OnClosedArgs(connection=<SelectConnection CLOSED socket=None params=<URLParameters host=localhost port=5672 virtual_host=/ ssl=False>>, reason_code=-1, reason_text='EOF')
Traceback (most recent call last):
  File "/Users/fokhunov/develop/python/build4use/parser/src/parser.py", line 1552, in <module>
    start()
  File "/Users/fokhunov/develop/python/build4use/parser/src/parser.py", line 1525, in start
    rmq_info("[ Tajikistan ] parsing started", tj_tag)
  File "/Users/fokhunov/develop/python/build4use/parser/src/parser.py", line 151, in rmq_info
    rqm_log(LL_INFO, msg, tag)
  File "/Users/fokhunov/develop/python/build4use/parser/src/parser.py", line 139, in rqm_log
    rmq_publish(rmq_route_logs, data)
  File "/Users/fokhunov/develop/python/build4use/parser/src/parser.py", line 124, in rmq_publish
    rmq_client.publish(route, json.dumps(data))
  File "/Users/fokhunov/develop/python/build4use/parser/src/producer.py", line 26, in publish
    self._channel.basic_publish(exchange=self._exchange, routing_key=route, body=data)
  File "/Library/Python/2.7/site-packages/pika/adapters/blocking_connection.py", line 2077, in basic_publish
    mandatory, immediate)
  File "/Library/Python/2.7/site-packages/pika/adapters/blocking_connection.py", line 2164, in publish
    self._flush_output()
  File "/Library/Python/2.7/site-packages/pika/adapters/blocking_connection.py", line 1250, in _flush_output
    *waiters)
  File "/Library/Python/2.7/site-packages/pika/adapters/blocking_connection.py", line 474, in _flush_output
    result.reason_text)
pika.exceptions.ConnectionClosed: (-1, 'EOF')
```

**Comment by lukebakken** [comment]
@fokhunov I have some questions -

* Can you provide all of the code you are using? It helps us a lot if we don't have to guess what code we need to write.
* Can you reproduce this outside of docker?
* What version of Erlang and RabbitMQ are used in the docker container?

**Comment by fokhunov** [comment]
@lukebakken 

1. I already posted my Producer code (see above). 
Here is some part of code (which still reproduces this issue).

```
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import argparse
import json
import signal
import sys
import time

import producer


print("__________________________________________")
print("Starting service %s" % time.strftime("%c"))

# general
service_name = "parser"

# arguments
parser = argparse.ArgumentParser(prog=service_name, description='parses web pages and calls API to collect exchange rates')
parser.add_argument('-rmq',      default='localhost',   help='rabbitmq host')
parser.add_argument('-interval', default='300',         help='parsing interval')
args = parser.parse_args()

# RabbitMQ
rmq_url = 'amqp://guest:guest@'+ args.rmq +':5672/%2F?heartbeat=0'
rmq_exchange = "erp"
rmq_route_logs = "logs"

rmq_client = producer.BlockingProducer(rmq_url, rmq_exchange)


def rqm_send(msg):
    data = {
        "service": service_name,
        "msg": msg
    }
    rmq_client.publish(rmq_route_logs, json.dumps(data))
    print(data)



"""
    Entry point to parser
"""


def finish(signal_num, stack_frame):
    print("Terminating service %s" % time.strftime("%c"))
    rqm_send("Terminating service %s" % time.strftime("%c"))
    rmq_client.close()
    sys.exit()


# handle Termination signal
signal.signal(signal.SIGTERM, finish)


if __name__ == "__main__":
    rqm_send("Starting service with " + str(args))
    counter = 0

    while True:
        rqm_send("Executing #" + str(counter))
        counter += 1
        time.sleep(float(args.interval))
```
2.  I tried to run my script with RabbitMQ which is not in Docker container. Result: everything works well. There are no connection closed, I waited more than 10 minutes for it.

Here is RabbitMQ status output:
```
/usr/local/sbin>rabbitmqctl status
Status of node rabbit@localhost ...
[{pid,11668},
 {running_applications,
     [{rabbitmq_stomp,"RabbitMQ STOMP plugin","3.7.5"},
      {rabbitmq_mqtt,"RabbitMQ MQTT Adapter","3.7.5"},
      {rabbitmq_management,"RabbitMQ Management Console","3.7.5"},
      {rabbitmq_web_dispatch,"RabbitMQ Web Dispatcher","3.7.5"},
      {rabbitmq_amqp1_0,"AMQP 1.0 support for RabbitMQ","3.7.5"},
      {amqp_client,"RabbitMQ AMQP Client","3.7.5"},
      {rabbitmq_management_agent,"RabbitMQ Management Agent","3.7.5"},
      {rabbit,"RabbitMQ","3.7.5"},
      {mnesia,"MNESIA  CXC 138 12","4.15.3"},
      {rabbit_common,
          "Modules shared by rabbitmq-server and rabbitmq-erlang-client",
          "3.7.5"},
      {ranch_proxy_protocol,"Ranch Proxy Protocol Transport","1.5.0"},
      {cowboy,"Small, fast, modern HTTP server.","2.2.2"},
      {ranch,"Socket acceptor pool for TCP protocols.","1.5.0"},
      {ssl,"Erlang/OTP SSL application","8.2.6"},
      {public_key,"Public key infrastructure","1.5.2"},
      {asn1,"The Erlang ASN1 compiler version 5.0.5","5.0.5"},
      {jsx,"a streaming, evented json parsing toolkit","2.8.2"},
      {amqp10_common,
          "Modules shared by rabbitmq-amqp1.0 and rabbitmq-amqp1.0-client",
          "3.7.5"},
      {cowlib,"Support library for manipulating Web protocols.","2.1.0"},
      {os_mon,"CPO  CXC 138 46","2.4.4"},
      {crypto,"CRYPTO","4.2.2"},
      {xmerl,"XML parser","1.3.16"},
      {recon,"Diagnostic tools for production use","2.3.2"},
      {sasl,"SASL  CXC 138 11","3.1.2"},
      {inets,"INETS  CXC 138 49","6.5.1"},
      {lager,"Erlang logging framework","3.5.1"},
      {goldrush,"Erlang event stream processor","0.1.9"},
      {compiler,"ERTS  CXC 138 10","7.1.5"},
      {syntax_tools,"Syntax tools","2.1.4"},
      {stdlib,"ERTS  CXC 138 10","3.4.5"},
      {kernel,"ERTS  CXC 138 10","5.4.3"}]},
 {os,{unix,darwin}},
 {erlang_version,
     "Erlang/OTP 20 [erts-9.3.1] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:64] [hipe] [kernel-poll:true] [dtrace]\n"},
 {memory,
     [{connection_readers,61896},
      {connection_writers,3120},
      {connection_channels,9624},
      {connection_other,135640},
      {queue_procs,17608},
      {queue_slave_procs,0},
      {plugins,2377352},
      {other_proc,22575424},
      {metrics,208280},
      {mgmt_db,236352},
      {mnesia,79152},
      {other_ets,2484328},
      {binary,1752464},
      {msg_index,30064},
      {code,29314811},
      {atom,1131721},
      {other_system,14086956},
      {allocated_unused,17671592},
      {reserved_unallocated,0},
      {strategy,rss},
      {total,[{erlang,74504792},{rss,59625472},{allocated,92176384}]}]},
 {alarms,[]},
 {listeners,
     [{clustering,25672,"::"},
      {amqp,5672,"127.0.0.1"},
      {http,15672,"::"},
      {mqtt,1883,"::"},
      {stomp,61613,"::"}]},
 {vm_memory_calculation_strategy,rss},
 {vm_memory_high_watermark,0.4},
 {vm_memory_limit,3435973836},
 {disk_free_limit,50000000},
 {disk_free,146002087936},
 {file_descriptors,
     [{total_limit,156},{total_used,4},{sockets_limit,138},{sockets_used,2}]},
 {processes,[{limit,1048576},{used,433}]},
 {run_queue,0},
 {uptime,976},
 {kernel,{net_ticktime,60}}]

```

3. This is from RabbitMQ container:

```
root@rabbitmq:/# rabbitmqctl status
Status of node rabbit@rabbitmq ...
[{pid,127},
 {running_applications,
     [{rabbitmq_management,"RabbitMQ Management Console","3.7.5"},
      {rabbitmq_management_agent,"RabbitMQ Management Agent","3.7.5"},
      {rabbitmq_web_dispatch,"RabbitMQ Web Dispatcher","3.7.5"},
      {rabbit,"RabbitMQ","3.7.5"},
      {amqp_client,"RabbitMQ AMQP Client","3.7.5"},
      {rabbit_common,
          "Modules shared by rabbitmq-server and rabbitmq-erlang-client",
          "3.7.5"},
      {ranch_proxy_protocol,"Ranch Proxy Protocol Transport","1.5.0"},
      {cowboy,"Small, fast, modern HTTP server.","2.2.2"},
      {ranch,"Socket acceptor pool for TCP protocols.","1.5.0"},
      {ssl,"Erlang/OTP SSL application","8.2.6"},
      {public_key,"Public key infrastructure","1.5.2"},
      {asn1,"The Erlang ASN1 compiler version 5.0.5","5.0.5"},
      {recon,"Diagnostic tools for production use","2.3.2"},
      {jsx,"a streaming, evented json parsing toolkit","2.8.2"},
      {os_mon,"CPO  CXC 138 46","2.4.4"},
      {cowlib,"Support library for manipulating Web protocols.","2.1.0"},
      {xmerl,"XML parser","1.3.16"},
      {crypto,"CRYPTO","4.2.1"},
      {inets,"INETS  CXC 138 49","6.5.1"},
      {mnesia,"MNESIA  CXC 138 12","4.15.3"},
      {lager,"Erlang logging framework","3.5.1"},
      {goldrush,"Erlang event stream processor","0.1.9"},
      {compiler,"ERTS  CXC 138 10","7.1.5"},
      {syntax_tools,"Syntax tools","2.1.4"},
      {sasl,"SASL  CXC 138 11","3.1.2"},
      {stdlib,"ERTS  CXC 138 10","3.4.5"},
      {kernel,"ERTS  CXC 138 10","5.4.3"}]},
 {os,{unix,linux}},
 {erlang_version,
     "Erlang/OTP 20 [erts-9.3.1] [source] [64-bit] [smp:2:2] [ds:2:2:10] [async-threads:64] [hipe] [kernel-poll:true]\n"},
 {memory,
     [{connection_readers,0},
      {connection_writers,0},
      {connection_channels,0},
      {connection_other,2840},
      {queue_procs,15752},
      {queue_slave_procs,0},
      {plugins,1780728},
      {other_proc,25547032},
      {metrics,197264},
      {mgmt_db,332040},
      {mnesia,74976},
      {other_ets,2174528},
      {binary,1215472},
      {msg_index,28848},
      {code,28437512},
      {atom,1123529},
      {other_system,22197343},
      {allocated_unused,21848520},
      {reserved_unallocated,0},
      {strategy,rss},
      {total,[{erlang,83127864},{rss,90521600},{allocated,104976384}]}]},
 {alarms,[]},
 {listeners,[{clustering,25672,"::"},{amqp,5672,"::"},{http,15672,"::"}]},
 {vm_memory_calculation_strategy,rss},
 {vm_memory_high_watermark,0.4},
 {vm_memory_limit,838426624},
 {disk_free_limit,50000000},
 {disk_free,57450815488},
 {file_descriptors,
     [{total_limit,1048476},
      {total_used,2},
      {sockets_limit,943626},
      {sockets_used,0}]},
 {processes,[{limit,1048576},{used,369}]},
 {run_queue,0},
 {uptime,64093},
 {kernel,{net_ticktime,60}}]

```

Any ideas ??

**Comment by fokhunov** [comment]
For sure problem in pika library. I tried run code in Golang and auto-close connection not happened.
Thanks.

**Comment by lukebakken** [comment]
> I already posted my Producer code (see above)

Hi @fokhunov - in general, it is best to provide running code that reproduces an issue when requested. I have to read and understand what you are doing before I can put together my own set of code to try to reproduce this issue. This takes away from time I could be using to get to the bottom of the issue.

I am planning on starting to look at this in a couple hours today (2018-06-04). If you have time to provide running code, that would be great, otherwise I will take time to distill what you have provided into something I can test with.

**Comment by lukebakken** [comment]
> I tried run code in Golang and auto-close connection not happened

Could you please provide the golang code you're using?

If you'd like me to try to reproduce this using Docker, please let me know **exactly** how to set up Docker to match your environment. Thanks.

**Comment by farrukh-okhunov** [comment]
@lukebakken Hi, 
1. Actually I provided all code. On my first post you can see `class Producer...` which I store in separate file and import it in main file as `import producer`.

2. For Go code I just used example from [RabbitMQ Go routing](https://www.rabbitmq.com/tutorials/tutorial-four-go.html)

3. My docker env:
```
# docker network create erp_network
# docker run -d --hostname rabbitmq --name rabbitmq --network erp_network -p 5672:5672 rabbitmq:3.7.5
# docker build -t parser .
# docker run -d --name parser --network erp_network -v /home/ec2-user/go/src/build4use/parser/src:/app parser -rmq=rabbitmq
```

Here is my app Dockerfile
```
FROM python:2.7.15-alpine3.7

# Copy and install dependencies
COPY requirements.txt /
RUN pip install -r /requirements.txt

# Set workdir. Script should be mounted.
WORKDIR /app

# Move app files inside image
COPY ./src/parser.py ./src/producer.py ./

# Run container
ENTRYPOINT [ "python", "-u", "parser.py" ]

# rabbitmq host based on environment (docker tip: use custom network)
# interval in seconds
CMD [ "-rmq=localhost", "-interval=300" ]

```

Here is my requirements.txt
```
beautifulsoup4==4.6.0
requests==2.18.4
pika==0.11.2
pytz==2013.7
```

That's all ;)

**Comment by lukebakken** [comment]
Thanks I will give all of that a try!

**Comment by lukebakken** [comment]
@farrukh-okhunov thanks for that information. I suspect the golang code runs successfully because it is not disabling heartbeats. I just now noticed that you are using `?heartbeat=0` - why is that? Disabling heartbeats is generally never a good idea. In addition, Pika `0.11.2` is affected by issue #1014 so RabbitMQ will still expect heartbeats from your application.

I have never had a need to use Docker so please pardon my new-user questions -

* I had to modify `producer.py` for it to work by adding `import pika` at the top. Is there anything else missing?
* How do you retrieve logs from your RabbitMQ docker image?
* If I create a `Dockerfile` for the python program I can see requirements installed and the container start up, however it does not remain started. What can I do to diagnose?
* If you re-run your test with heartbeats enabled, do you see the same issue?
* Could you please test using version [`0.12.0b3`](https://pypi.org/project/pika/0.12.0b3/)? This version includes a fix for issue #1014 which will tell RabbitMQ to disable heartbeats if the value is set to `0`.

**Comment by lukebakken** [comment]
@farrukh-okhunov I am able to reproduce what you describe without using docker.

Please see the changes I made in [this commit](https://github.com/lukebakken/pika-1053/commit/85af0dcb48ba542bbc41ee8daf130731e2636de0) which shows how to ensure heartbeats are processed while using `BlockingConnection`.

In your code, you should enable heartbeats, and use `process_data_events()` periodically to ensure they are processed. Or, you could use the asynchronous publisher code [here](https://github.com/pika/pika/blob/0.12.0b3/examples/asynchronous_publisher_example.py) as a starting point, as it will periodically send heartbeats.

Here is a summary of what I think is affecting you:

* You attempt to disable heartbeats but they are not actually disabled due to the bug described in #1014. This is fixed in `0.12.0b3`, so I would like you to test with that version.
* Rather than disabling heartbeats, you should either use `SelectConnection` or modify your code in a manner to what I have done [here](https://github.com/lukebakken/pika-1053/commit/85af0dcb48ba542bbc41ee8daf130731e2636de0) to ensure that heartbeats are processed.
* You did not see this issue using the golang code as it will process heartbeats asynchronously by default.

I am going to close this issue now but please feel free to comment and re-open it if you have further questions or find something new. Thanks again for patiently providing the information I requested.
